### PR TITLE
model instantiation outside hydra

### PR DESCRIPTION
**Pull request by okuchaiev**
Making sure self._cfg object of model is of DictConfig type.

### CODE REVIEW

**Comment by okuchaiev** [comment]
@titu1994 still this solution doesn't seem to work for dataclasses (see not passing tests/core/test_asr_ctcencdec_model.py test).

@titu1994 , @tkornuta-nvidia @ericharper any other ideas on how to convert dataclass (with nested DictConfig or dataclass members) to DictConfig?

File: examples/asr/speech_to_text.py [hunk 1]
```diff
@@ -65,9 +64,7 @@
 
 @hydra.main(config_path="conf", config_name="config")
 def main(cfg):
-    # omegaconf merg trainer stuff to optim - this is necessary to be able to correctly setup LR scheduler
     trainer = pl.Trainer(**cfg.pl.trainer)
-    logging.info(f'Hydra config: {cfg.pretty()}')
```
Comment (titu1994) [inline examples/asr/speech_to_text.py:70]: Perhaps we should log this now inside ModelPT, after creation of self._cfg?

File: nemo/core/classes/modelPT.py [hunk 2]
```diff
@@ -46,9 +45,29 @@ class ModelPT(LightningModule, Model):
...
+                f"cfg argument must be of type ModelPTconfig, DictConfig or Dict. But got {type(cfg)} instead."
+            )
+        if isinstance(cfg, DictConfig):
+            self._cfg = cfg
+        else:
+            if dataclasses.is_dataclass(cfg):
+                cfg = dataclasses.asdict(cfg)
+
+            self._cfg = OmegaConf.create(cfg)
         self.save_hyperparameters(self._cfg)
```
Comment (titu1994) [inline nemo/core/classes/modelPT.py:71]: Important Note: We need to remove this `self.save_hyperparameters(self._cfg)` line. The LightningModel.hparams is used for logging more than anything - and neither Tensorflow nor other loggers support nested dictionaries for logging. On the other hand, pushing the immutable DictConfig inside this means that we cannot update / inject train time parameters - lr, loss value etc since these keys dont exist inside self._cfg anymore. Even if we down cast to dict each iteration, push these values and then convert back to DictConfig (horribly expensive operation per iteration), we cannot actually export this since during deserialization, it would find wrong keys like "lr" and "train_loss" etc.

So please remove this and directly use self._cfg in order to serialize / deserialize.

File: examples/asr/speech_to_text.py (hunk 1 above)
Comment (tkornuta-nvidia) [reply examples/asr/speech_to_text.py:70]: Not sure if we should log this at all as:
1) hydra is saving the "resulting" configuration in a separate yaml file
https://hydra.cc/docs/tutorial/working_directory
2) user has few command line args (`--h`, `--cfg` ) to display the configuration

Comment (okuchaiev) [reply examples/asr/speech_to_text.py:70]: let's not log it in ModelPT

File: nemo/core/classes/modelPT.py (hunk 2 above)
Comment (tkornuta-nvidia) [reply nemo/core/classes/modelPT.py:71]: hmmmm, @okuchaiev please correct me, but this was not for serialization/deserialization.

If I understood that added you added this as wanted to display params per a given experiment in tensorboard - and it worked.

Not sure how it will handle e.g. schedulers changing learning rate though...

Comment (okuchaiev) [reply nemo/core/classes/modelPT.py:71]: discussed

**Comment by okuchaiev** [comment]
this is really now just adjusting asr to DictConfig
